src/PostgREST/App.hs

```diff
@@ -84,23 +84,48 @@ postgrest logLev refConf refDbStructure pool getTime connWorker =
               case jwtClaims attempt of
                 Left errJwt -> return . errorResponseFor $ errJwt
                 Right claims -> do
-                  let
-                    authed = containsRole claims
-                    shouldCommit   = configDbTxAllowOverride conf && iPreferTransaction apiRequest == Just Commit
-                    shouldRollback = configDbTxAllowOverride conf && iPreferTransaction apiRequest == Just Rollback
-                    preferenceApplied
-                      | shouldCommit    = addHeadersIfNotIncluded [(hPreferenceApplied, BS.pack (show Commit))]
-                      | shouldRollback  = addHeadersIfNotIncluded [(hPreferenceApplied, BS.pack (show Rollback))]
-                      | otherwise       = identity
-                    handleReq = do
-                      when (shouldRollback || (configDbTxRollbackAll conf && not shouldCommit)) HT.condemn
-                      mapResponseHeaders preferenceApplied <$> runPgLocals conf claims (app dbStructure conf) apiRequest
-                  dbResp <- P.use pool $ HT.transaction HT.ReadCommitted (txMode apiRequest) handleReq
-                  return $ either (errorResponseFor . PgError authed) identity dbResp
+                  dbResp <- P.use pool .
+                    HT.transaction HT.ReadCommitted (txMode apiRequest) .
+                    optionalRollback conf apiRequest $
+                    runPgLocals conf claims (app dbStructure conf) apiRequest
+                  return $ either (errorResponseFor . PgError (containsRole claims)) identity dbResp
         -- Launch the connWorker when the connection is down. The postgrest function can respond successfully(with a stale schema cache) before the connWorker is done.
         when (responseStatus response == status503) connWorker
         respond response
 
+optionalRollback
+  :: AppConfig
+  -> ApiRequest
+  -> HT.Transaction Response
+  -> HT.Transaction Response
+optionalRollback conf apiRequest tx =
```

Comment:
I refactored the `configDbTxEnd` on top of the `optionalRollback` function from https://github.com/PostgREST/postgrest/pull/1725. I updated it with the TxEnd constructors.

I could remove this commit and leave it for a later PR..